File: matlab/test/dmp_kf_test/sim_DMP_Pos_EKF_disc.m
function sim_DMP_Pos_EKF_disc()

%% ==============================================================
%% DMP with state reset (y=y_r, y_dot=y_r_dot) and force feedback, i.e. 
%% y_ddot = h(theta, y_r, y_r_dot, t) + f_ext/M

% clc;
% close all;
% clear;

format compact;
set_matlab_utils_path();

%% ###################################################################

setPosParams();

path = strrep(mfilename('fullpath'), 'sim_DMP_Pos_EKF_disc','');

load([path 'data/dmp_data.mat'],'dmp_p', 'Yg0', 'Y0', 'tau0');

Dim = length(Y0);
can_clock_ptr = dmp_p.getCanClockPtr();
can_clock_ptr.setTau(tau0);

%% DMP simulation
% set initial values
t = 0.0;
iters = 0;
x = 0.0; dx = 0.0;
Y0 = Y0 + Y0_offset;
Y = Y0; dY = zeros(Dim,1); ddY = zeros(Dim,1);
F_ext = zeros(Dim,1);

Time = [];
Y_data = []; dY_data = []; ddY_data = [];
x_data = [];
Yg_data = [];
tau_data = [];
F_data = [];
Sigma_theta_data = [];

Yg = Yg0 + Yg_offset;

t_end = tau0 + time_offset;
tau = t_end;
can_clock_ptr.setTau(tau);

tau_hat = tau0;
Yg_hat = Yg0;
if (~est_goal), Yg_hat = Yg; end
if (~est_tau), tau_hat = tau; end
x_hat = t/tau_hat;

N_out = length(Yg_hat);
Y_out_hat = zeros(N_out,1);
Y_out = zeros(N_out,1);

theta = [];
if (est_goal),theta = Yg_hat; end
if (est_tau), theta = [theta; tau_hat]; end

N_params = length(theta);
P_theta = eye(N_params, N_params) .* init_params_variance;
Rn_hat = eye(N_out,N_out) .* msr_noise_hat;
Qn = eye(N_params,N_params) .* process_noise;

rng(0);
Rn = eye(N_out,N_out) .* msr_noise;
Sigma_vn = sqrt(Rn);


%% Set up EKF object
ekf = EKF(N_params, N_out, @pStateTransFun, @pMsrFun);
ekf.setProcessNoiseCov(Qn);
ekf.setMeasureNoiseCov(Rn_hat);
% a_p = exp(a_pc*dt);
ekf.setFadingMemoryCoeff(a_p);
ekf.theta = theta;
ekf.P = P_theta;

ekf.enableParamsContraints(enable_constraints);
ekf.setParamsConstraints(A_c, b_c);
ekf.setPartDerivStep(num_diff_step);

ekf.setStateTransFunJacob(@pStateTransFunJacob);
% ekf.setMsrFunJacob(@pMsrFunJacob);

dmp_p.setY0(Y0);

disp('DMP-EKF (discrete) Pos simulation...');
tic
while (true)

    %% data logging

    Time = [Time t];

    Y_data = [Y_data Y];
    dY_data = [dY_data dY];  
    ddY_data = [ddY_data ddY];

    Yg_data = [Yg_data Yg_hat];
    tau_data = [tau_data tau_hat];

    x_data = [x_data x];

    F_data = [F_data F_ext];

    Sigma_theta_data = [Sigma_theta_data sqrt(diag(P_theta))];

    %% DMP simulation  
    ddY = dmp_p.calcYddot(x, Y, dY, Yg) + Sigma_vn*randn(N_out,1);
    ddY_hat = dmp_p.calcYddot(x_hat, Y, dY, Yg_hat);

    F_ext = M_r * (ddY - ddY_hat);
    
    Y_out = ddY;
    % Y_out_hat = ddY_hat;

    %% Update phase variable
    dx = can_clock_ptr.getPhaseDot(x);

    %% Stopping criteria
    err_p = norm(Yg-Y)/norm(Yg);
    if (err_p <= 0.5e-2 && t>=t_end)
        break; 
    end

    iters = iters + 1;
    if (t>=t_end)
        warning('Time limit reached. Stopping simulation...\n');
        break;
    end

    %% ========  KF measurement update  ========
    ekf.correct(Y_out, pMsrCookie(dmp_p, t, Y, dY, Yg, tau, est_goal, est_tau));
    
    %% ========  KF time update  ========
    ekf.predict();

    theta = ekf.theta;
    P_theta = ekf.P;

    %% Numerical integration
    t = t + dt;
    x = x + dx*dt;
    Y = Y + dY*dt;
    dY = dY + ddY*dt;
    
    if (est_goal), Yg_hat = theta(1:end-est_tau); end
    if (est_tau), tau_hat = theta(end); end;
    x_hat = t/tau_hat;

end
toc

figure
hold on;
plot(Time, Sigma_theta_data(1,:), 'LineWidth',2.0, 'Color','red');
plot(Time, Sigma_theta_data(2,:), 'LineWidth',2.0, 'Color','green');
plot(Time, Sigma_theta_data(3,:), 'LineWidth',2.0, 'Color','blue');
plot(Time, Sigma_theta_data(4,:), 'LineWidth',2.0, 'Color',[0.93 0.69 0.13]);
legend({'$x$','$y$','$z$','$\tau$'}, 'interpreter','latex', 'fontsize',15);
title('$\sigma_{\theta}$', 'interpreter','latex', 'fontsize',15);
xlabel('time [$s$]', 'interpreter','latex', 'fontsize',15);
hold off;

plot_pos_estimation_results(Time, Yg, Yg_data, tau, tau_data, Sigma_theta_data, F_data, plot_1sigma, Y_data, dY_data);

end

function theta_next = pStateTransFun(theta, cookie)
    
    theta_next = theta;
    
end

function J = pStateTransFunJacob(theta, cookie)
    
    n_params = length(theta);
    J = eye(n_params, n_params);
    
end

function Y_out = pMsrFun(theta, cookie)

    if (cookie.est_goal), Yg_hat = theta(1:end-cookie.est_tau);
    else, Yg_hat = cookie.Yg;
    end
    
    if (cookie.est_tau), tau_hat = theta(end);
    else, tau_hat = cookie.tau;
    end

    x_hat = cookie.t/tau_hat;
    
    Y_out = cookie.dmp_p.calcYddot(x_hat, cookie.Y, cookie.dY, Yg_hat);

end

function J = pMsrFunJacob(theta, cookie)
    
    if (cookie.est_goal), Yg_hat = theta(1:end-cookie.est_tau);
    else, Yg_hat = cookie.Yg;
    end
    
    if (cookie.est_tau), tau_hat = theta(end);
    else, tau_hat = cookie.tau;
    end
    
    x_hat = cookie.t / tau_hat;

    J = cookie.dmp_p.getAcellPartDev_g_tau(cookie.t, cookie.Y, cookie.dY, x_hat, Yg_hat, tau_hat);

end

function cookie = pMsrCookie(dmp_p, t, Y, dY, Yg, tau, est_goal, est_tau)
    
    cookie = struct('dmp_p',dmp_p, 't',t, 'Y',Y, 'dY',dY, 'Yg',Yg, 'tau',tau, 'est_goal',est_goal, 'est_tau',est_tau);
    
end

File: c++/src/test/dmp_kf_test/src/sim_DMP_Pos_EKF_disc.cpp
#include <ros/ros.h>
#include <ros/package.h>
#include <memory>
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <exception>

#include <io_lib/io_utils.h>
#include <dmp_lib/DMP/DMP_pos.h>
#include <dmp_lib/GatingFunction/LinGatingFunction.h>
#include <dmp_lib/GatingFunction/ExpGatingFunction.h>
#include <dmp_lib/GatingFunction/SigmoidGatingFunction.h>
#include <kf_lib/EKF.h>

#include <dmp_kf_test/utils.h>

using namespace as64_;

struct PosMsrCookie
{
  std::shared_ptr<dmp_::DMP_pos> dmp_p;
  double t;
  arma::vec Y;
  arma::vec dY;
  arma::vec Y0;
  arma::vec Yg;
  double tau;

  PosMsrCookie(std::shared_ptr<dmp_::DMP_pos> dmp_p, double t,
               const arma::vec &Y, const arma::vec &dY, const arma::vec &Y0,
               const arma::vec &Yg, double tau)
  {
    this->dmp_p = dmp_p;
    this->t = t;
    this->Y = Y;
    this->dY = dY;
    this->Y0 = Y0;
    this->Yg = Yg;
    this->tau = tau;
  }
};

arma::vec pStateTransFun(const arma::vec &theta, void *cookie)
{
  return theta;
}

arma::vec pMsrFun(const arma::vec &theta, void *cookie)
{
  PosMsrCookie *ck = static_cast<PosMsrCookie *>(cookie);

  arma::vec Yg_hat = theta.subvec(0,2);
  double tau_hat = theta(3);

  double x_hat = ck->t/tau_hat;

  ck->dmp_p->setTau(tau_hat);
  arma::vec Y_out = ck->dmp_p->calcYddot(x_hat, ck->Y, ck->dY, Yg_hat);
  ck->dmp_p->setTau(ck->tau);

  return Y_out;
}


int main(int argc, char** argv)
{
  // ===========  Initialize the ROS node  ===============
  ros::init(argc, argv, "dmp_ekf_pos_test_node");
  ros::NodeHandle nh_("~");

  std::string path = ros::package::getPath("dmp_kf_test");

  double dt = 0.005;

  arma::vec Y0_offset = {0.0, 0.0, 0.0};

  arma::vec Yg_offset = {0.75, -0.09, 0.64};
  double time_offset = 6;

  arma::vec Yg_up_lim = 0.8*arma::vec({1.0, 1.0, 1.0,});
  arma::vec Yg_low_lim = -Yg_up_lim;
  double tau_low_lim = 1.0;
  double tau_up_lim = 30.0; // Inf;

  arma::mat process_noise = arma::diagmat(arma::vec({0.02, 0.02, 0.02, 0.05})); // Q
  double msr_noise = std::pow(0.2,2);
  double msr_noise_hat = 1000;


  arma::mat init_params_variance = arma::diagmat( 1e5*arma::vec({1, 1, 1, 1}) ); // P
  double a_pc = 0.99; // forgetting factor in fading memory KF
  double a_p = 1.002;

  arma::vec theta_low_lim = arma::join_vert(Yg_low_lim, arma::vec({tau_low_lim}) );
  arma::vec theta_up_lim = arma::join_vert(Yg_up_lim, arma::vec({tau_up_lim}) );

  int N_params = theta_low_lim.size();
  arma::mat A_c = arma::join_vert(-arma::mat().eye(N_params, N_params), arma::mat().eye(N_params, N_params));
  arma::vec b_c = arma::join_vert(-theta_low_lim, theta_up_lim);
  bool enable_constraints = true;

  arma::vec num_diff_step = {0.001, 0.001, 0.001, 0.01};

  arma::mat M_r = 5*arma::mat().eye(3,3);

  // ===========  load DMP data  ===============
  std::string dmp_data_file = path + "/matlab/data/dmp_data.bin";
  std::shared_ptr<dmp_::DMP_pos> dmp_p;
  std::shared_ptr<dmp_::DMP_eo> dmp_o;
  arma::vec P0d;
  arma::vec Yg0;
  arma::vec Q0d;
  arma::vec Qgd;
  double tau0;
  loadDMPdata(dmp_data_file, &dmp_p, &dmp_o, &Yg0, &P0d, &Qgd, &Q0d, &tau0);

  int Dim = Yg0.size();
  std::shared_ptr<dmp_::CanonicalClock> can_clock_ptr = dmp_p->can_clock_ptr;
  can_clock_ptr->setTau(tau0);

  // DMP simulation
  // set initial values
  double t = 0.0;
  double x = 0.0;
  double dx = 0.0;
  arma::vec Y0 = P0d + Y0_offset;
  arma::vec Y = Y0;
  arma::vec dY = arma::vec().zeros(Dim);
  arma::vec ddY = arma::vec().zeros(Dim);
  arma::vec F_ext = arma::vec().zeros(Dim);

  arma::rowvec Time;
  arma::mat Y_data;
  arma::mat dY_data;
  arma::mat ddY_data;
  arma::rowvec x_data;
  arma::mat Yg_data;
  arma::rowvec tau_data;
  arma::mat F_data;
  arma::mat Sigma_theta_data;

  arma::vec Yg = Yg0 + Yg_offset;

  double t_end = tau0 + time_offset;
  double tau = t_end;
  can_clock_ptr->setTau(tau);

  double tau_hat = tau0;
  arma::vec Yg_hat = Yg0;
  double x_hat = t/tau_hat;

  int N_out = Yg_hat.size();
  arma::vec Y_out_hat = arma::vec().zeros(N_out);
  arma::vec Y_out = arma::vec().zeros(N_out);

  arma::vec theta = arma::join_vert(Yg_hat, arma::vec({tau_hat}));

  arma::mat P_theta = arma::mat().eye(N_params, N_params) % init_params_variance;
  arma::mat Rn_hat = arma::mat().eye(N_out,N_out) * msr_noise_hat;
  arma::mat Qn = arma::mat().eye(N_params,N_params) % process_noise;

  arma::arma_rng::set_seed(0);
  arma::mat Rn = arma::mat().eye(N_out,N_out) * msr_noise;
  arma::mat Sigma_vn = arma::sqrtmat_sympd(Rn);

  // Set up EKF object
  kf_::EKF ekf(theta, P_theta, N_out, &pStateTransFun, &pMsrFun);
  ekf.setProcessNoiseCov(Qn);
  ekf.setMeasureNoiseCov(Rn_hat);
  // a_p = exp(a_pc*dt);
  ekf.setFadingMemoryCoeff(a_p);

  ekf.enableParamsContraints(enable_constraints);
  ekf.setParamsConstraints(A_c, b_c);
  ekf.setPartDerivStep(num_diff_step);

  // ekf.setStateTransFunJacob(@pStateTransFunJacob);
  // ekf.setMsrFunJacob(@pMsrFunJacob);

  dmp_p->setY0(Y0);

  std::cout << "DMP-EKF (discrete) Pos simulation...\n";
  arma::wall_clock timer;
  timer.tic();
  while (true)
  {
    // data logging
    Time = arma::join_horiz(Time, arma::vec({t}));
    Y_data = arma::join_horiz(Y_data, Y);
    dY_data = arma::join_horiz(dY_data, dY);
    ddY_data = arma::join_horiz(ddY_data, ddY);

    Yg_data = arma::join_horiz(Yg_data, Yg_hat);
    tau_data = arma::join_horiz(tau_data, arma::vec({tau_hat}));

    x_data = arma::join_horiz(x_data, arma::vec({x}));
    F_data = arma::join_horiz(F_data, F_ext);
    Sigma_theta_data = arma::join_horiz(Sigma_theta_data, arma::sqrt(arma::diagvec(P_theta)));

    // DMP simulation
    dmp_p->setTau(tau_hat);
    arma::vec ddY_hat = dmp_p->calcYddot(x_hat, Y, dY, Yg_hat);
    dmp_p->setTau(tau);

    ddY = dmp_p->calcYddot(x, Y, dY, Yg) + Sigma_vn*arma::vec().randn(N_out);


    F_ext = M_r * (ddY - ddY_hat);

    Y_out = ddY;
    // Y_out_hat = ddY_hat;

    // Update phase variable
    dx = can_clock_ptr->getPhaseDot(x);

    // Stopping criteria
    double err_p = arma::norm(Yg-Y)/arma::norm(Yg);
    if (err_p <= 0.5e-2 & t>=t_end) break;

    if (t>=t_end)
    {
      std::cerr << "Time limit reached. Stopping simulation...\n";
      break;
    }

    // ========  KF measurement update  ========
    PosMsrCookie msr_cookie(dmp_p, t, Y, dY, Y0, Yg, tau);
    ekf.correct(Y_out, static_cast<void *>(&msr_cookie));

    // ========  KF time update  ========
    ekf.predict();

    theta = ekf.theta;
    P_theta = ekf.P;

    // Numerical integration
    t = t + dt;
    x = x + dx*dt;
    Y = Y + dY*dt;
    dY = dY + ddY*dt;

    Yg_hat = theta.subvec(0,2);
    tau_hat = theta(3);
    x_hat = t/tau_hat;

  }
  std::cout << "Elapsed time: " << timer.toc() << " sec\n";

  // ===========  write results  ===============
  std::string sim_data_file = path + "/matlab/data/pos_est_results.bin";
  std::ofstream out(sim_data_file, std::ios::out | std::ios::binary);
  if (!out) throw std::runtime_error("Failed to create file \"" + sim_data_file + "\"...");
  io_::write_mat(Time, out);
  io_::write_mat(Yg, out);
  io_::write_mat(Yg_data, out);
  io_::write_scalar(tau, out);
  io_::write_mat(tau_data, out);
  io_::write_mat(Sigma_theta_data, out);
  io_::write_mat(F_data, out);
  io_::write_mat(Y_data, out);
  io_::write_mat(dY_data, out);
  out.close();

  // ===========  Shutdown ROS node  ==================
  ros::shutdown();

  return 0;
}
